dace/transformation/passes/reference_reduction.py

```diff
@@ -0,0 +1,247 @@
+        self.remove_refsets(result, refsets)
+
+        # Reconnect reference uses as views
+        self.reconnect_views(sdfg, result, access_states, reference_sources)
+
+        # Modify data descriptor from Reference to View
+        self.change_ref_descriptors_to_views(sdfg, result)
+
+        return result or None
+
+    def report(self, pass_retval: Set[str]) -> str:
+        return f'Converted {len(pass_retval)} references to views: {pass_retval}.'
+
+    def find_refsets(self, candidates: Set[str],
+                     access_states: Dict[str, Set[SDFGState]]) -> Dict[str, List[Tuple[SDFGState, nodes.AccessNode]]]:
+        """
+        Returns a dictionary of reference name to a list of tuples of (state, access node)
+        where the reference is set via a memlet.
+        """
+        result: Dict[str, List[Tuple[SDFGState, nodes.AccessNode]]] = defaultdict(list)
+        all_states_to_consider: Set[SDFGState] = set()
+        for candidate in candidates:
+            all_states_to_consider.update(access_states[candidate])
+
+        for state in all_states_to_consider:
+            # Loop over all states that use the references once
+            for node in state.data_nodes():
+                if node.data not in candidates:
+                    continue
+                for _ in state.in_edges_by_connector(node, 'set'):
+                    result[node.data].append((state, node))
+                    break
+
+        return result
+
+    def find_candidates(
+        self,
+        sdfg: SDFG,
+        reference_sources: Dict[str, Set[Memlet]],
+        refsets: Dict[str, List[Tuple[SDFGState, nodes.AccessNode]]],
+        access_states: Dict[str, Set[SDFGState]],
+        reachable_states: Dict[SDFGState, Set[SDFGState]],
+    ) -> Set[str]:
+        """
+        Returns a set of candidates for conversion to views.
+        """
+        result = set(refsets.keys())
+        if not result:  # Early return
+            return result
+
+        # If memlet does not depend on any symbol, it can be kept. Otherwise,
+        # it may depend on a (free) symbol. There are multiple options:
+        #   * If dependent on scope symbol (e.g., map parameter) - remove from candidates
+        #   * If dependent on symbol defined in inter-state edges - make sure it is not changed between set and uses
+        #   * If dependent on a free symbol - also make sure it is not changed between set and uses
+        for cand in list(result):  # Copying the set to a list allows us to iterate over it while removing elements
+            source = next(iter(reference_sources[cand]))
+            fsyms = source.subset.free_symbols
+            if not fsyms:
+                continue
+
+            for state, node in refsets[cand]:
+                # Check if any of the symbols is a scope symbol
+                entry = state.entry_node(node)
+                while entry is not None:
+                    if fsyms & entry.new_symbols(sdfg, state, {}):
```

Comment:
Lines 124-129 are not hit by the tests, if it's easy to add one that causes these edge cases then it may be good to add one. Otherwise nevermind.